adapters/amqp-vertx/src/main/java/org/eclipse/hono/adapter/amqp/VertxBasedAmqpProtocolAdapter.java

```diff
@@ -287,71 +287,69 @@ private void handleSessionOpen(final ProtonConnection conn, final ProtonSession
+
+            receiver.setTarget(null);
+            receiver.setQoS(receiver.getRemoteQoS());
+            if (ProtonQoS.AT_LEAST_ONCE.equals(receiver.getRemoteQoS())) {
+                // disable auto-accept for this transfer model.
+                // in this case, the adapter will apply the required disposition
+                receiver.setAutoAccept(false);
+            }
+            receiver.handler((delivery, message) -> {
+                validateMessageAddress(message.getAddress(), authenticatedDevice)
+                .compose(validAddress -> {
+                    message.setAddress(validAddress.toString());
+                    uploadMessage(new AmqpContext(delivery, message, authenticatedDevice));
+                    return Future.succeededFuture();
+                }).recover(t -> {
+                    // invalid message address
+                    MessageHelper.rejected(delivery, AmqpContext.getErrorCondition(t));
+                    return Future.failedFuture(t);
                 });
-                HonoProtonHelper.setCloseHandler(receiver, remoteDetach -> onLinkDetach(receiver));
-                receiver.open();
-                return Future.succeededFuture();
             });
+            HonoProtonHelper.setCloseHandler(receiver, remoteDetach -> onLinkDetach(receiver));
+            receiver.open();
         }
     }
 
     /**
      * Forwards a message received from a device to downstream consumers.
      * 
+     * @param qos The QoS level of the recever link.
      * @param context The context that the message has been received in.
      */
     protected void uploadMessage(final AmqpContext context) {
-        final Future<Void> formalCheck = Future.future();
+        final Future<Void> contentTypeCheck = Future.future();
         final String contentType = context.getMessageContentType();
 
-        if (isPayloadOfIndicatedType(context.getMessagePayload(), contentType)) {
-            if (!context.isDeviceAuthenticated() && context.getDeviceId() == null) {
-                formalCheck.fail(new ClientErrorException(HttpURLConnection.HTTP_BAD_REQUEST,
-                        String.format("Missing device-id in [address: %s] for unauthenticated device", context.getResourceIdentifier())));
-            } else {
-                formalCheck.complete();
-            }
-        } else {
-            formalCheck.fail(new ClientErrorException(HttpURLConnection.HTTP_BAD_REQUEST,
+        if (!isPayloadOfIndicatedType(context.getMessagePayload(), contentType)) {
+            contentTypeCheck.fail(new ClientErrorException(HttpURLConnection.HTTP_BAD_REQUEST,
                     String.format("Content-Type: [%s] does not match payload", contentType)));
+        } else {
+            contentTypeCheck.complete();
         }
 
-        formalCheck.compose(ok -> {
+        CompositeFuture.all(contentTypeCheck, validateEndpoint(context)).compose(ok -> {
```

Comment:
is there a reason why you want to validate the endpoint as part of the uploadMessage method instead of doing it as part of the message address check like the MQTT adapter does it? It helps people to understand the different adapters more easily if we use similar patterns in the adapters. So, unless there are any compelling reasons to do otherwise, IMHO you should follow the pattern used in the MQTT adapter, i.e. check the endpoint (and QoS) of the incoming message and then validate the message's address before handing it over to the uploadMessage method. WDYT?

Reply:
the problem is that validating the endpoint requires a context instance which is already created after validating the address. let me see how to refactor that